gtsam/linear/PowerMethod.h

```diff
@@ -46,94 +46,72 @@ class PowerMethod {
 
   size_t nrIterations_;  // number of iterations
 
-  double ritzValue_;   // all Ritz eigenvalues
-  Vector ritzVector_;  // all Ritz eigenvectors
+  double ritzValue_;   // Ritz eigenvalue
+  Vector ritzVector_;  // Ritz eigenvector
 
  public:
   // Constructor
   explicit PowerMethod(const Operator &A,
                        const boost::optional<Vector> initial = boost::none)
       : A_(A), dim_(A.rows()), nrIterations_(0) {
     Vector x0;
-    x0 = initial ? Vector::Random(dim_) : initial.get();
+    x0 = initial ? initial.get() : Vector::Random(dim_);
     x0.normalize();
 
-    // initialize Ritz eigen values
+    // initialize Ritz eigen value
     ritzValue_ = 0.0;
 
-    // initialize Ritz eigen vectors
+    // initialize Ritz eigen vector
     ritzVector_ = Vector::Zero(dim_);
-
-    ritzVector_.col(0) = update(x0);
-    perturb();
+    ritzVector_ = powerIteration(x0);
   }
 
-  // Update the vector by dot product with A_
-  Vector update(const Vector &x) const {
+  // Update the vector by dot product with A_, and return A * x / || A * x ||
+  Vector powerIteration(const Vector &x) const {
     Vector y = A_ * x;
     y.normalize();
     return y;
   }
 
-  // Update the vector by dot product with A_
-  Vector update() const { return update(ritzVector_); }
-
-  // Perturb the initial ritzvector
-  void perturb() {
-    // generate a 0.03*||x_0||_2 as stated in David's paper
-    std::mt19937 rng(42);
-    std::uniform_real_distribution<double> uniform01(0.0, 1.0);
-
-    int n = dim_;
-    // Vector disturb(n);
-    // for (int i = 0; i < n; ++i) {
-    //   disturb(i) = uniform01(rng);
-    // }
-    Vector disturb = Vector::Random(n);
-    disturb.normalize();
-
-    Vector x0 = ritzVector_;
-    double magnitude = x0.norm();
-    ritzVector_ = x0 + 0.03 * magnitude * disturb;
-  }
+  //  Update the vector by dot product with A_, and return A * x / || A * x ||
+  Vector powerIteration() const { return powerIteration(ritzVector_); }
 
-  // Perform power iteration on a single Ritz value
-  // Updates ritzValue_
-  bool iterateOne(double tol) {
+  // After Perform power iteration on a single Ritz value, if the error is less
+  // than the tol then return true else false
+  bool converged(double tol) {
     const Vector x = ritzVector_;
-    double theta = x.transpose() * A_ * x;
-
     // store the Ritz eigen value
-    ritzValue_ = theta;
-
-    const Vector diff = A_ * x - theta * x;
-    double error = diff.norm();
+    ritzValue_ = x.dot(A_ * x);
```

Comment:
This line is **changing** ```ritzValue_```  -- simple test functions like this should **not** cause side-effects!

You should change this to be a temporary value, and make this function ```const```.